refactor!: new state var traits + Owned

benesjan opened the pull request:
Fixes F-189

In this PR I:
- Introduce new `StateVariable` trait that replaces the `HasStorageSlot` trait. The new trait has a `new` method on it that allowed me to significantly clean up the `#[storage]` macro - now I can directly construct the state var by calling this method instead of doing some crazy stuff with closures etc.,
- I introduce `OwnedStateVariable` trait that is used by state variables that have the concept of owner (`PrivateSet`, `PrivateMutable` etc.). These state vars need to be wrapped in new `Owned` state var before being placed in the `Storage` struct.

benesjan commented:
* **#18792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18815** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18815?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18782** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18782?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18721** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18721?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18721?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/storage.nr:
```diff
@@ -135,62 +146,3 @@ pub comptime fn storage_no_init(s: TypeDefinition) {
-    (constructor, get_storage_size(typ))
-}
-
-/// Returns true if `typ` is `state_vars::map::Map`.
-comptime fn is_storage_map(typ: Type) -> bool {
-    if typ.as_data_type().is_some() {
-        let (def, generics) = typ.as_data_type().unwrap();
-        let maybe_map = if (def.name() == quote { Map }) & (generics.len() == 3) {
-            let maybe_key = generics[0];
-            let maybe_value = generics[1];
-            let maybe_context = generics[2];
-            quote { crate::state_vars::map::Map<$maybe_key, $maybe_value, $maybe_context> }.as_type()
-        } else {
-            quote {()}.as_type()
-        };
-        typ == maybe_map
-    } else {
-        false
-    }
-}
-
-comptime fn add_context_generic(typ: Type, context_generic: Type) -> Type {
-    let (def, mut generics) = typ.as_data_type().expect(
-        f"Storage containers must be generic structs of the form `Container<..., Context>`",
-    );
-    let name = def.name();
-
-    if is_storage_map(typ) {
-        generics[generics.len() - 2] = add_context_generic(generics[1], context_generic);
-        generics[generics.len() - 1] = context_generic;
-    } else {
-        generics[generics.len() - 1] = context_generic;
-    }
-
-    let generics = generics.map(|typ: Type| quote {$typ}).join(quote {,});
-    quote { $name<$generics> }.as_type()
-}
```
This is currently unused but would be relevant for implementing https://github.com/AztecProtocol/aztec-packages/issues/8658. Put there a [comment](https://github.com/AztecProtocol/aztec-packages/issues/8658#issuecomment-3608951047) such that this doesn't get lost if it's ever needed again.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/storage.nr:
```diff
@@ -29,56 +31,59 @@ pub comptime fn storage(s: TypeDefinition) -> Quoted {
+    let mut storage_layout_struct_members = &[];
     let mut storage_layout_constructors = &[];
 
-    // TODO(#8658): uncomment the code below to inject the Context type parameter.
-    //let mut new_storage_fields = &[];
-    //let context_generic = s.add_generic("Context");
```
See comment at the end of the file on why I think it's fine to drop this mess from here.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/storage.nr:
```diff
@@ -2,17 +2,19 @@ use poseidon::poseidon2::Poseidon2Hasher;
 use super::utils::AsStrQuote;
-use super::utils::get_storage_size;
 
 /// Stores a map from a module to the name of the struct that describes its storage layout.
 /// This is then used when generating a `storage_layout()` getter on the contract struct.
 pub comptime mut global STORAGE_LAYOUT_NAME: UHashMap<Module, Quoted, BuildHasherDefault<Poseidon2Hasher>> =
     UHashMap::default();
```
I would say it's most likely possible to drop this and instead iterate over structs in the `#[aztec]` macro and generate the functionality there. Similarly to what we did with function macros.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/storage.nr:
```diff
@@ -29,56 +31,59 @@ pub comptime fn storage(s: TypeDefinition) -> Quoted {
 
-    //s.set_fields(new_storage_fields);
     let storage_vars_constructors = storage_vars_constructors.join(quote {,});
-    let storage_impl = quote {
-        impl<Context> Storage<Context> {
-            fn init(context: Context) -> Self {
-                Self {
-                    $storage_vars_constructors
-                }
-            }
-        }
-    };
```
Moved this to the final quote as it's cleaner.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/storage.nr:
```diff
@@ -135,62 +146,3 @@ pub comptime fn storage_no_init(s: TypeDefinition) {
-        quote { $struct_name::new(context, $slot)}
-    };
-
-    (constructor, get_storage_size(typ))
```
So much spaghetti. Happy it's gone.

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/owned_state_variable.nr:
```diff
@@ -0,0 +1,38 @@
+/// 2. has the ability to nullify (spend/destroy) those notes.
+///
+/// (Note that point 1 does not necessarily have to be true; for example, the note message could be encrypted for a
+/// decrypting service address, which could then deliver the note to the actual owner via offchain communication.)
```
It felt hard to explain why Owned is generally to be used only with private data and it feels like we need to coin some new terms here.

Particularly I am not sure if this is a good thing to say:
> in the public context, ownership does not exist.

because I imagine that people would think that ownership applies also to public data and that a person could own that public data if a contract gives him permissions to modify it.

WDYT? Should we genuinely introduce something like nelly? The owner here feels quite overloaded.

mverzilli replied:
I understand the concern you're raising but also I feel the explanation you wrote above is super clear. We're not talking about ownership in broad terms but specifically about _ownership of state variables_, and from that perspective the concept of Owner seems quite well defined to me. 

Re: nelly, introducing a new term is tempting because then we could just say "it's a nelly ☝️", but also comes at the expense of explaining the term across the board. So my vote here would be for `Owned` over some new word.

Maybe we could point someone from the community or devrel to these lines and tell us what they think/what they understand/what they intuit from it? Just to get a fresher perspective

mverzilli replied:
> in the public context, ownership does not exist.

Regarding this, maybe we can just do without it? 

"Owned state variables are only used when working with private data because, in the public context, write permissions are solely defined by the contract logic, and anyone can read data."

benesjan replied:
I like

https://github.com/AztecProtocol/aztec-packages/pull/18721/commits/18087777a4e66c4679e7693525a2dd90228d43f9

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_immutable.nr:
```diff
@@ -137,36 +83,15 @@ impl<Note> OwnedPrivateImmutable<Note, &mut PrivateContext> {
         self.compute_initialization_nullifier(secret)
     }
 
```
Pardon me for I have sinned and unnecessarily increased the scope of this PR. As once discussed in scrum the rigid style of the docs below is just too verbose and the `Advanced` section is describing implementation which is unhelpful.

Since I was already touching the docs decided to make it more concise and formatted it to 120 char line width.

iAmMichaelConnor replied:
My beautiful comments! 😱 
The advanced sections were for the benefit of advanced folks -- based on the feedback we got from a certain very advanced user who once wrote a big thread on slack to say it wasn't clear what these functions do behind the scenes.
RIP to the beautiful 100-character lines for comments 😭

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/storage.nr:
```diff
@@ -1,8 +1,5 @@
-    fn get_storage_slot(self) -> Field;
-}
+// TODO(benesjan): This is just for the macro-generated StorageLayout struct. It seems strange to have it here and
+// the file needs renaming to storable.nr.
```
The `state_vars` dir is now quite messy (would group owned and normal state vars in separate dirs) so would tackle this in a followup PR.

mverzilli commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -90,18 +90,25 @@ where
 
 `PrivateImmutable`, `PrivateMutable` and `PrivateSet` got modified to directly contain the owner instead of implicitly "containing it" by including it in the storage slot via a `Map`.
-Now the `Map` is redundant and the relevant state variable should be moved out of it:
+These state variable now implement a newly introduced `OwnedStateVariable` trait (see docs of `OwnedStateVariable` for explanation of what it is).
```
```suggestion
These state variables now implement a newly introduced `OwnedStateVariable` trait (see docs of `OwnedStateVariable` for explanation of what it is).
```

mverzilli commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -90,18 +90,25 @@ where
 `PrivateImmutable`, `PrivateMutable` and `PrivateSet` got modified to directly contain the owner instead of implicitly "containing it" by including it in the storage slot via a `Map`.
-Now the `Map` is redundant and the relevant state variable should be moved out of it:
+These state variable now implement a newly introduced `OwnedStateVariable` trait (see docs of `OwnedStateVariable` for explanation of what it is).
+This changes makes the state variables incompatible with `Map` and now instead these should be wrapped in new `Owned` state variable:
```
```suggestion
These changes make the state variables incompatible with `Map` and now instead these should be wrapped in new `Owned` state variable:
```

mverzilli commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -90,18 +90,25 @@ where
 struct Storage<Context> {
 -    private_nfts: Map<AztecAddress, PrivateSet<NFTNote, Context>, Context>,
-+    private_nfts: PrivateSet<NFTNote, Context>,
++    private_nfts: Owned<PrivateSet<NFTNote, Context>, Context>,
```
cool!

mverzilli commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -90,18 +90,25 @@ where
 
-Now we have an `at` method on the private state variables that scopes it to a given owner.
-Given that we used to call `at` on the map and now we call it directly on the set if you so the above change it should not be required of you to do any further changes in your contract.
+If you used your state variable with a `Map` that had `AztecAddress` type as key then the usage is unchanged:
```
I find it difficult to parse this sentence, suggestion:


```suggestion
Note that even though the types of your state variables are changing from `Map<AztecAddress, T, Context>` to `Owned<T, Context>`, usage remains unchanged:
```

mverzilli commented on docs/docs/developers/docs/resources/migration_notes.md:
````diff
@@ -90,18 +90,25 @@ where
+```
+
+With this change the underlying notes will inherit the storage slot of the `Owned` state variable.
+This is unlike in `Map` where the nested state variable got the storage slot computed as `hash([map_storage_slot, key])`.
````
```suggestion
This is unlike `Map` where the nested state variable got the storage slot computed as `hash([map_storage_slot, key])`.
```

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -20,137 +20,73 @@ use protocol_types::{
-///
-/// You can declare a state variable of type PrivateMutable within your contract's
-/// #[storage] struct:
+/// PrivateMutable is an owned state variable type represents a private value that can be changed. Because it is
```
```suggestion
/// PrivateMutable is an owned state variable type that represents a private value that can be changed. Because it is
```

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -20,137 +20,73 @@ use protocol_types::{
-/// the state variable has exactly one value at any point in time. However, the
-/// underlying implementation differs significantly because of Aztec's private
-/// state model.
+/// For more details on what "owned" means, see the documentation for the OwnedStateVariable trait.
```
I think `OwnedStateVariable` can be turned into a doc link

benesjan replied:
Was not aware of that feature. Very cool. Thanks

https://github.com/AztecProtocol/aztec-packages/pull/18721/commits/f354e0bd73ecb508d4a9afa29534ae56dcd9892f

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -20,137 +20,73 @@ use protocol_types::{
-/// returns an OwnedPrivateMutable handle for that owner.
+/// This is conceptually similar to how a regular variable works in Ethereum: the state variable has exactly one value
+/// at any point in time. However, the underlying implementation differs significantly because of Aztec's private state
+/// model.
```
With the decluttering work you've done here, now this feels redundant and too detailed: we already say that the variable is mutable above, and we point the reader to learn more about implementation details at OwnedStateVariable. 

The reference to similarities and differences between Ethereum and Aztec's private state model is a bit orphaned without the actual explanation of what happens with notes, so IMO it's better to just kill this paragraph

benesjan replied:
Agree, we were explaining private state again and again in the docs so I nuked these comments:

https://github.com/AztecProtocol/aztec-packages/pull/18721/commits/20f54e5bfb5b51bbaa11b8d694f049e07c75e310

They all felt unhelpful.

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -20,137 +20,73 @@ use protocol_types::{
-/// A users' account nonce can be represented as a PrivateMutable<NonceNote>.
-/// The "current value" of the user's nonce is the value contained within the
-/// single not-yet-nullified note in the user's view of the PrivateMutable.
+/// A users' account nonce can be represented as a PrivateMutable<NonceNote>. The "current value" of the user's nonce is
```
```suggestion
/// A user's account nonce can be represented as a PrivateMutable<NonceNote>. The "current value" of the user's nonce is
```

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -20,137 +20,73 @@ use protocol_types::{
-/// - Use PrivateSet when you want to allow "other people" (beyond the owner) to
-///   insert notes into the state variable.
+/// - Use PrivateMutable when you want exactly one note to represent the state variable's current value, similar to
+///   regular variables in Ethereum.
```
Should we say "similar to regular variables in Solidity" instead?

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18721/commits/218a2f18b4b9e3caacf48dab7f873c7280f14377

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -20,137 +20,73 @@ use protocol_types::{
-/// designed to not leak anything about _which_ state variable was read/modified/
-/// initialized, to the outside world.
+/// The methods of a PrivateMutable are only executable in a PrivateContext, and are designed to not leak anything about
+/// _which_ state variable was read/modified/ initialized, to the outside world.
```
```suggestion
/// _which_ state variable was read/modified/initialized, to the outside world.
```

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/private_set.nr:
```diff
@@ -11,63 +11,47 @@ use crate::{
-/// You can declare a state variable of type PrivateSet within your contract's
-/// #[storage] struct:
+/// PrivateSet is an owned state variable type, which enables you to read, mutate, and write private state. Because it
+/// is "owned," you must wrap a PrivateImmutable inside an Owned state variable when storing it:
```
Maybe a `PrivateImmutable` slipped here?

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18721/commits/baa457d6ca7ca73b1b29a1c7f35fbaf31e8cc1b8

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/state_variable.nr:
```diff
@@ -0,0 +1,23 @@
+/// A trait that defines the common interface for all Aztec state variables. State variables are variables that can be
+/// stored in the contract's storage. Examples of state variables are `Owned`, `PublicMutable`, `Map` or
+/// `DelayedPublicMutable`.
+///
+/// # Type Parameters
+///
+/// * `N` - A compile-time constant that determines how many storage slots need to be reserved for this state variable.
+/// * `Context` - The execution context type (e.g., `PublicContext`, `PrivateContext`, `UtilityContext`). The context
+///               determines which methods of the state variable are available and controls whether the variable can
+///               be accessed in public, private or utility functions.
+///
+pub trait StateVariable<let N: u32, Context> {
+    /// Initializes a new state variable at a given storage slot.
+    ///
+    /// This function is automatically called within the #[storage] macro and needs to be manually called only when
+    /// #[storage_no_init] is used.
+    ///
+    fn new(context: Context, storage_slot: Field) -> Self;
+
+    /// Returns the storage slot at which the state variable is placed. Typically used only when testing the contract or
+    /// when using a lower-level API like `get_notes` function.
+    fn get_storage_slot(self) -> Field;
+}
```
Love how this turned out!

mverzilli reviewed (commented):
Super neat! I left a ton of comments mostly around documentation but most of them are nitpicks. I don't dare approve this because I don't have enough insight, but it looks really good from my humble perspective

mverzilli reviewed (approved):
🚢

benesjan committed:
refactor!: new state var traits + Owned

Fixes F-189

In this PR I:
- Introduce new `StateVariable` trait that replaces the `HasStorageSlot` trait. The new trait has a `new` method on it that allowed me to significantly clean up the `#[storage]` macro - now I can directly construct the state var by calling this method instead of doing some crazy stuff with closures etc.,
- I introduce `OwnedStateVariable` trait that is used by state variables that have the concept of owner (`PrivateSet`, `PrivateMutable` etc.). These state vars need to be wrapped in new `Owned` state var before being placed in the `Storage` struct.

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
